bpf/lib/nodeport.h

```diff
@@ -327,20 +327,41 @@ static __always_inline int find_dsr_v6(struct __ctx_buff *ctx, __u8 nexthdr,
 
 static __always_inline int
 nodeport_extract_dsr_v6(struct __ctx_buff *ctx, struct ipv6hdr *ip6,
+			struct ipv6_ct_tuple *tuple, int l4_off,
 			union v6addr *addr, __be16 *port, bool *dsr)
 {
 	struct dsr_opt_v6 opt __align_stack_8 = {};
+	struct ipv6_ct_tuple tmp = *tuple;
 	int ret;
 
+	if (tuple->nexthdr == IPPROTO_TCP) {
+		union tcp_flags tcp_flags = {};
+
+		if (l4_load_tcp_flags(ctx, l4_off, &tcp_flags) < 0)
+			return DROP_CT_INVALID_HDR;
+
+		ipv6_ct_tuple_reverse(&tmp);
+
+		if (!(tcp_flags.value & TCP_FLAG_SYN)) {
+			*dsr = ct_has_dsr_egress_entry6(get_ct_map6(&tmp), &tmp);
+			*port = 0;
+			return 0;
+		}
+	}
+
 	ret = find_dsr_v6(ctx, ip6->nexthdr, &opt, dsr);
 	if (ret != 0)
 		return ret;
 
 	if (*dsr) {
 		*addr = opt.addr;
 		*port = opt.port;
+		return 0;
 	}
 
+	if (tuple->nexthdr == IPPROTO_TCP)
+		ct_update_dsr(get_ct_map6(&tmp), &tmp, false);
```

Comment:
Why is this needed? Isn't the DSR flag set to false already in the CT entry? I thought that's what we just read `*dsr` from.

Reply:
Oh, I had not realized the `.dsr` flag was only kept for the TCP SYN packet of each connection and removed afterward. That's new with those changes, right?

Reply:
Ah no, let me elaborate a bit more.

For TCP traffic, `nodeport_extract_dsr_v4()` does three things:
1. for non-SYNs, check if they belong to a DSR connection (== match a DSR CT entry). If so, send the packet towards the DSR_INGRESS code (without DSR info).
2. for SYNs, extract DSR info from the IP option. If the info exists, send the packet towards the DSR_INGRESS code (with DSR info).
3. that leaves us with SYNs which don't have embedded DSR info. So effectively a fresh, non-DSR connection. But the same Client->BackendIP connection might have been **previously** established as DSR connection, and still have a dangling DSR CT entry. If we don't clean up that CT entry now, all subsequent non-SYN packets of this fresh (non-DSR!) connection will end up matching that stale CT entry (step 1) and believe they belong to a DSR connection.

Reply:
Got it! Thanks for the explanation!

I don't know if it's new or if I missed it before, but the current comment in `nodeport_extract_dsr_v4` for that code looks good enough to me.